allen-cell-animated/nucmorph-colorizer
# feat: Categorical thresholding

Problem
=======
Continues work on #148, implementing thresholding on categorical features.

Solution
========
This change is large enough that I'm planning to do a code walkthrough with any interested parties in the new year!

- Updates `ColorizeCanvas` to filter using both categorical and numeric thresholds.
- Updates the `FeatureThreshold` type to handle both categorical and numeric (continuous/discrete number) features.

URL Changes:
- `url_utils.ts`: Adds support for serializing/deserializing categorical thresholds to the URL. Selected categorical thresholds are saved as hex strings.

UI Changes:
- Rendering for categorical features in the thresholds panel.
- Adds scroll shadows to the thresholds panel, based on designs from Lyndsay.
- Fixes an inconsistent layout bug with the side panel when browser width changes.

## Type of change
* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open the preview link: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-170/?dataset=https%3A%2F%2Fraw.githubusercontent.com%2Fallen-cell-animated%2Fnucmorph-colorizer%2Ffeature%2Fcategorical-thresholding%2Fdocumentation%2Fexample_dataset%2Fmanifest.json&feature=Continuous%20Feature&t=4&color=matplotlib-cool
2. Go to the filters tab and select one or more categorical features.
3. Change the Color map at the top. The checkboxes should update.
4. Enable/disable checkboxes and watch the matching objects be enabled/disabled in the main viewport.
5. Copy the URL and open a new tab. Your filters should remain constant.

Screenshots (optional):
-----------------------

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/4848a710-0fd2-4b05-bbd8-2c8d624609cf)

Video Talkthrough (🔊)

https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/5f9e9ebf-8cbc-4f73-acdf-1b5883a9b05b


Keyfiles (delete if not relevant):
-----------------------
1. `src/colorizer/types.ts` -> changes to `FeatureThreshold` type
6. `ColorizeCanvas.ts` -> filtering with the new `FeatureThresold` categorical type
7. `Dataset.ts` -> minor API change, `tryGetFeatureData`
8. `App.tsx`
9. `FeatureThresholdsTab.tsx` -> rendering of thresholds

Thanks for contributing!
